File: Python/367.Valid Perfect Square.py
#判断一个数是否是完全平方数，不能用开方等运算。
class Solution(object):

    #二分法，在1和Num/2之间找平方为Num的数
    def isPerfectSquare(self, num):
        L, R = 1, (num >> 1) + 1
        while L <= R:
            m = L + ((R - L) >> 1)
            mul = m ** 2
            if mul == num:
                return True
            elif mul > num:
                R = m - 1
            else:
                L = m + 1
        return False

    #蛮力法，一个一个找
    def isPerfectSquare1(self, num):
        n , i = 1 , 1
        while n<=num:
            n = i*i
            if n==num:
                return True
            elif n>num:
                return False
            i += 1

s = Solution()
print(s.isPerfectSquare(100))

File: C++/Binary Search/367.Valid Perfect Square.cpp
/*
判断一个数是否是完全平方数。
*/
/*
思路：使用二分法计算平方是否大于给定数。
 */

#include <iostream>
using namespace std;

class Solution {
public:
	bool isPerfectSquare(int num) {
		int left = 1, right = num;
		while(left <= right)
		{
			long mid = left + (right - left)/2;// long类型防止2147483647用例
			long square = mid*mid;
			if(square == num)	return true;
			else if(square > num)	right = mid-1;
			else left = mid+1;
		}
		return false;
    }
};

int main()
{
	Solution s;
	cout<<s.isPerfectSquare(256)<<endl;
}

